src/shared/validators.ts

```diff
@@ -0,0 +1,207 @@
+import { UInt4 } from 'bitwise/types';
+import { MAX_ADDR, encReqCodes, EMPTY_WALLET_UID, ASCII_REGEX } from '../constants';
+import { isUInt4, checksum } from '../util';
+
+export const validateValueExists = (arg: { [key: string]: any }) => {
+  const [key, [, value]] = Object.entries(arg);
+  if (!value) {
+    throw new Error(`${key} must be provided`);
+  }
+};
+
+export const validateIsUInt4 = (n?: number) => {
+  if (typeof n !== 'number' || !isUInt4(n)) {
+    throw new Error('Must be an integer between 0 and 15 inclusive');
+  }
+  return n as UInt4;
+};
+
+export const validateNAddresses = (n: number) => {
+  validateIsUInt4(n);
+  if (n > MAX_ADDR)
+    throw new Error(`You may only request ${MAX_ADDR} addresses at once.`);
+};
+
+export const validateStartPath = (startPath: number[]) => {
+  if (!startPath) {
+    throw new Error('Start path is required');
+  }
+  if (startPath.length < 2 || startPath.length > 5)
+    throw new Error('Path must include between 2 and 5 indices');
+};
+
+export const validateDeviceId = (deviceId?: string) => {
+  if (!deviceId) {
+    throw new Error(
+      'No device ID has been stored. Please connect with your device ID first.',
+    );
+  }
+  return deviceId;
+};
+
+export const validateEncryptRequestCode = (code: keyof typeof encReqCodes) => {
+  if (code && encReqCodes[code] === undefined) {
+    throw new Error('Unknown encrypted request code.');
+  }
+};
+
+export const validateAppName = (name?: string) => {
+  if (!name) {
+    throw new Error('Name is required.');
+  }
+  if (name.length < 5 || name.length > 24) {
+    throw new Error(
+      'Invalid length for name provided. Must be 5-24 characters.',
+    );
+  }
+  return name;
+};
+
+export const validateResponse = (res: Buffer) => {
+  if (!res) {
+    throw new Error('Error decrypting response');
+  }
+};
+
+export const validateUrl = (url?: string) => {
+  if (!url) {
+    throw new Error('Url does not exist. Please reconnect.');
+  }
+  return url;
+};
+
+export const validateBaseUrl = (baseUrl?: string) => {
+  if (!baseUrl) {
+    throw new Error('Base URL is required.');
+  }
+  return baseUrl;
+};
+
+export const validateFwConstants = (fwConstants?: FirmwareConstants) => {
+  if (!fwConstants) {
+    throw new Error('Firmware constants do not exist. Please reconnect.');
+  }
+  return fwConstants;
+};
+export const validateFwVersion = (fwVersion?: Buffer) => {
+  if (!fwVersion) {
```

Comment:
You could also check that it is 3 bytes

Reply:
I am seeing 4 bytes back from the backend. We could just say >4 bytes is an error?

Reply:
Oh right! sorry

Reply:
991ebaa

Reply:
Huh, never used `byteLength` before